File: Bachelor/Semester4/Systems_for_Design_and_Implementation/C#/ProiectCShMPP/Networking/Utils/RpcConcurrentServer.cs
using System.Net.Sockets;
using System.Threading;
using Networking.RpcProtocol;
using Services;

namespace Networking.Utils
{
    public class RpcConcurrentServer : AbstractConcurrentServer
    {
        private readonly IService _service;
        
        public RpcConcurrentServer(string ip, int port, IService service) : base(ip, port)
        {
            _service = service;
        }

        protected override Thread CreateWorker(TcpClient client)
        {
            return new Thread(new ClientRpcWorker(_service, client).Run);
        }
    }
}

File: Bachelor/Semester5/Parallel_and_Distributed_Programming/P1/Networking/src/main/java/utils/RpcConcurrentServer.java
package utils;

import rpcprotocol.ClientRpcWorker;
import services.Service;

import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class RpcConcurrentServer extends AbstractConcurrentServer {
    private final Service service;
    private final ExecutorService executorService;
    private ScheduledExecutorService scheduledExecutorService;
    private static final int NR_SECONDS_TO_CHECK = 5;

    public RpcConcurrentServer(int port, Service service, int nrThreads) {
        super(port);
        this.service = service;
        this.executorService = Executors.newFixedThreadPool(nrThreads);
    }

    @Override
    protected void createWorker(Socket client) {
        executorService.execute(new ClientRpcWorker(service, client));
    }

    @Override
    protected void scheduleChecks() {
        scheduledExecutorService = Executors.newScheduledThreadPool(1);
        scheduledExecutorService.scheduleAtFixedRate(service::checkShows, 0, NR_SECONDS_TO_CHECK, TimeUnit.SECONDS);
    }

    @Override
    public void stop() {
        super.stop();
        service.notifyObserversServerCloses();
        executorService.shutdown();
        scheduledExecutorService.shutdown();
    }
}